### PR TITLE
Enable nullability in ListView

### PR DESCRIPTION
**Pull request by gpetrou**
## Proposed changes

- Enable nullability in ListView.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6627)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ItemCheckedEventArgs.cs [hunk 1]
```diff
@@ -9,14 +9,14 @@ namespace System.Windows.Forms
     /// </summary>
     public class ItemCheckedEventArgs : EventArgs
     {
-        public ItemCheckedEventArgs(ListViewItem? item)
+        public ItemCheckedEventArgs(ListViewItem item)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ItemCheckedEventArgs.cs:12]: We must validate this in this case, e.g. `.ThrowIfNull()`.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 2]
```diff
@@ -2463,15 +2454,15 @@ private void CheckDisplayIndices()
...
             {
                 return;
             }
 
             IO.FileInfo fi;
-            for (int i = 0; i <= bkImgFileNamesCount; i++)
+            for (int i = 0; i <= _bkImgFileNamesCount; i++)
             {
-                fi = new IO.FileInfo(bkImgFileNames[i]);
+                fi = new IO.FileInfo(_bkImgFileNames[i]!);
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs]: We can't assume it's not null (see `SetBackgroundImage()` method). We should null check, and continue.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 3]
```diff
@@ -2768,7 +2759,7 @@ unsafe void CustomDraw(ref Message m)
                                 if (nmcd->iSubItem == 0 && Items[itemIndex].SubItems.Count > 1)
                                 {
                                     // Use the width for the first column header.
-                                    subItemBounds.Width = columnHeaders[0].Width;
+                                    subItemBounds.Width = _columnHeaders![0].Width;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs]: We can't assume this. In other parts of the class we always null-check before dereferencing the value.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 4]
```diff
@@ -3169,14 +3159,14 @@ protected override void Dispose(bool disposing)
...
                         catch (IO.IOException) { }
-                        backgroundImageFileName = string.Empty;
+                        _backgroundImageFileName = string.Empty;
                     }
 
-                    for (int i = 0; i <= bkImgFileNamesCount; i++)
+                    for (int i = 0; i <= _bkImgFileNamesCount; i++)
                     {
-                        fi = new IO.FileInfo(bkImgFileNames[i]);
+                        fi = new IO.FileInfo(_bkImgFileNames![i]!);
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs]: As above

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 5]
```diff
@@ -5387,19 +5368,19 @@ private unsafe void SetDisplayIndices(int[] indices)
         /// </summary>
         internal void UpdateSavedCheckedItems(ListViewItem item, bool addItem)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs]: Propose to refactor as follows:
```cs
            if (addItem)
            {
                savedCheckedItems ??= new List<ListViewItem>();
                savedCheckedItems.Add(item);
            }
            else if (savedCheckedItems is not null)
            {
                Debug.Assert(savedCheckedItems.Contains(item), "somehow we lost track of one item");
                savedCheckedItems.Remove(item);
            }
```

**Review by RussKie — CHANGES_REQUESTED** [review]
In the future it'd really appreciated it if you presented non-functional changes (e.g., variable renames, code clean up, etc.) in separate commits from functional changes (such as NRTA), as it'd greatly improve the readability of the changeset.
Thanks

**Comment by gpetrou** [comment]
> In the future it'd really appreciated it if you presented non-functional changes (e.g., variable renames, code clean up, etc.) in separate commits from functional changes (such as NRTA), as it'd greatly improve the readability of the changeset. Thanks

Sure. I'll keep that in mind.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 6]
```diff
@@ -2459,22 +2459,25 @@ private void CleanPreviousBackgroundImageFiles()
...
             }
 
-            IO.FileInfo fi;
+            FileInfo fi;
             for (int i = 0; i <= _bkImgFileNamesCount; i++)
             {
-                fi = new IO.FileInfo(_bkImgFileNames[i]!);
-                if (fi.Exists)
+                var bkImgFileName = _bkImgFileNames[i];
+                if (bkImgFileName is not null)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs:2466]: Please invert
```cs
if (bkImgFileName is null)
{
    continue;
}
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 7]
```diff
@@ -3158,24 +3164,31 @@ protected override void Dispose(bool disposing)
...
-                        // ComCtl ListView uses COM objects to manipulate the bitmap we send it to them.
-                        // I could not find any resources which explain in detail when the IImgCtx objects
-                        // release the temporary file. So if we get a FileIO when we delete the temporary file
-                        // we don't do anything about it ( because we don't know what is a good time to try to delete the file again ).
-                        try
+                        if (_bkImgFileNames is not null)
                         {
-                            fi.Delete();
+                            var bkImgFileName = _bkImgFileNames[i];
+                            if (bkImgFileName is not null)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs:3176]: ditto

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 8]
```diff
@@ -99,112 +97,111 @@ public partial class ListView : Control
         private const int MAXTILECOLUMNS = 20;
 
         // PERF: take all the bools and put them into a state variable
-        private Collections.Specialized.BitVector32 listViewState; // see LISTVIEWSTATE_ consts above
-        private Collections.Specialized.BitVector32 listViewState1; // see LISTVIEWSTATE1_ consts above
+        private Collections.Specialized.BitVector32 _listViewState; // see LISTVIEWSTATE_ consts above
```
[the remark this answers is not in the document]
Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs]: Doesn't seem like they can.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 9]
```diff
@@ -1810,10 +1803,10 @@ public ListViewItem TopItem
                         Scroll(0, value.Index);                     // To work around this, we force the listbox to display the first item, then scroll down to the item
                     }                                               // user is setting as the top item.
 
-                    return;                                         //
-                }                                                   //
+                    return;
+                }
 
-                if (value.Index == topItem.Index)
+                if (value.Index == topItem!.Index)
```
[the remark this answers is not in the document]
Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs]: Sorry, not sure what you mean here? Should I change something?

[the remark this answers is not in the document]
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs:1809]: We can't assume `topItem` is not null here. I think we should just return if `topItem` is null.

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs (hunk 5 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs:5380]: This is still outstanding.

Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs:5380]: Hmm, I see that I already applied the proposed change. Not sure why you don't see it.
image

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs:5380]: Apologies, you have indeed. I must have looked at the stale diff...

**Review by RussKie — APPROVED** [review]

**Comment by gpetrou** [comment]
@RussKie perhaps we could merge this?

**Comment by RussKie** [comment]
I can indeed. Thank you for the nudge.

**Pull request merged**